File: src/main/java/gg/mineral/practice/PracticePlugin.java
package gg.mineral.practice;

import org.bukkit.Bukkit;
import org.bukkit.command.Command;
import org.bukkit.event.Listener;
import org.bukkit.generator.ChunkGenerator;
import org.bukkit.plugin.java.JavaPlugin;

import gg.mineral.practice.commands.config.ArenaCommand;
import gg.mineral.practice.commands.config.CatagoryCommand;
import gg.mineral.practice.commands.config.GametypeCommand;
import gg.mineral.practice.commands.config.KitEditorCommand;
import gg.mineral.practice.commands.config.LeaderboardConfigCommand;
import gg.mineral.practice.commands.config.ListConfigCommands;
import gg.mineral.practice.commands.config.LobbyCommand;
import gg.mineral.practice.commands.config.PartiesCommand;
import gg.mineral.practice.commands.config.QueuetypeCommand;
import gg.mineral.practice.commands.config.SettingsConfigCommand;
import gg.mineral.practice.commands.config.SpectateConfigCommand;
import gg.mineral.practice.commands.duel.AcceptCommand;
import gg.mineral.practice.commands.duel.DuelCommand;
import gg.mineral.practice.commands.events.EventCommand;
import gg.mineral.practice.commands.kit.LeaveCommand;
import gg.mineral.practice.commands.party.PartyCommand;
import gg.mineral.practice.commands.settings.DayCommand;
import gg.mineral.practice.commands.settings.NightCommand;
import gg.mineral.practice.commands.settings.SettingsCommand;
import gg.mineral.practice.commands.settings.ToggleDuelRequestsCommand;
import gg.mineral.practice.commands.settings.TogglePartyRequestsCommand;
import gg.mineral.practice.commands.settings.TogglePlayerVisibilityCommand;
import gg.mineral.practice.commands.settings.ToggleScoreboardCommand;
import gg.mineral.practice.commands.spectator.FollowCommand;
import gg.mineral.practice.commands.spectator.SpectateCommand;
import gg.mineral.practice.commands.spectator.StopSpectatingCommand;
import gg.mineral.practice.commands.stats.EloCommand;
import gg.mineral.practice.commands.stats.LeaderboardsCommand;
import gg.mineral.practice.commands.stats.PotsCommand;
import gg.mineral.practice.commands.stats.ViewInventoryCommand;
import gg.mineral.practice.commands.testing.BotTestingCommand;
import gg.mineral.practice.commands.tournament.JoinCommand;
import gg.mineral.practice.commands.tournament.TournamentCommand;
import gg.mineral.practice.listeners.BuildListener;
import gg.mineral.practice.listeners.CommandListener;
import gg.mineral.practice.listeners.ComsumeListener;
import gg.mineral.practice.listeners.DamageListener;
import gg.mineral.practice.listeners.DeathListener;
import gg.mineral.practice.listeners.EntryListener;
import gg.mineral.practice.listeners.HealthListener;
import gg.mineral.practice.listeners.InteractListener;
import gg.mineral.practice.listeners.InventoryListener;
import gg.mineral.practice.listeners.MovementListener;
import gg.mineral.practice.listeners.PacketListener;
import gg.mineral.practice.listeners.ProjectileListener;
import gg.mineral.practice.managers.ArenaManager;
import gg.mineral.practice.managers.CatagoryManager;
import gg.mineral.practice.managers.GametypeManager;
import gg.mineral.practice.managers.KitEditorManager;
import gg.mineral.practice.managers.LeaderboardManager;
import gg.mineral.practice.managers.PartyManager;
import gg.mineral.practice.managers.PlayerSettingsManager;
import gg.mineral.practice.managers.ProfileManager;
import gg.mineral.practice.managers.QueuetypeManager;
import gg.mineral.practice.managers.SpectateManager;
import gg.mineral.practice.util.world.VoidWorldGenerator;
import lombok.val;
import net.minecraft.server.v1_8_R3.MinecraftServer;

public class PracticePlugin extends JavaPlugin {

	public static PracticePlugin INSTANCE;

	@Override
	public void onEnable() {

		INSTANCE = this;

		ProfileManager.load();
		PlayerSettingsManager.load();
		SpectateManager.load();
		LeaderboardManager.load();
		PartyManager.load();
		ArenaManager.load();
		QueuetypeManager.load();
		KitEditorManager.load();
		PartyManager.load();
		PlayerSettingsManager.load();
		CatagoryManager.load();
		GametypeManager.load();

		registerCommands(new ListConfigCommands(), new ArenaCommand(), new QueuetypeCommand(), new GametypeCommand(),
				new KitEditorCommand(), new LobbyCommand(), new PartiesCommand(),
				new AcceptCommand(), new ViewInventoryCommand(), new DuelCommand(), new SpectateCommand(),
				new PotsCommand(), new EloCommand(), new LeaderboardsCommand(), new PartyCommand(),
				new SettingsConfigCommand(), new FollowCommand(), new TogglePlayerVisibilityCommand(),
				new ToggleDuelRequestsCommand(), new TogglePartyRequestsCommand(), new CatagoryCommand(),
				new SettingsCommand(),
				new StopSpectatingCommand(), new TournamentCommand(), new JoinCommand(), new EventCommand(),
				new LeaveCommand(), new SpectateConfigCommand(),
				new LeaderboardConfigCommand(), new ToggleScoreboardCommand(), new BotTestingCommand(),
				new DayCommand(), new NightCommand());

		registerListeners(new BuildListener(), new InteractListener(), new ComsumeListener(), new InventoryListener(),
				new DeathListener(), new DamageListener(), new EntryListener(), new HealthListener(),
				new MovementListener(), new ProjectileListener(), new CommandListener(), new PacketListener());
	}

	@Override
	public void onDisable() {
		Bukkit.getServer().getScheduler().cancelTasks(this);
	}

	@Override
	public ChunkGenerator getDefaultWorldGenerator(String worldName, String id) {
		return new VoidWorldGenerator();
	}

	public void registerCommands(Command... cmds) {
		for (val cmd : cmds)
			MinecraftServer.getServer().server.getCommandMap().registerOverride(cmd.getName(), "Practice", cmd);
	}

	public void registerListeners(Listener... listeners) {
		for (val listener : listeners)
			Bukkit.getPluginManager().registerEvents(listener, this);
	}
}

File: src/main/kotlin/gg/mineral/practice/PracticePlugin.kt
package gg.mineral.practice

import com.github.retrooper.packetevents.PacketEvents
import com.github.retrooper.packetevents.event.PacketListenerPriority
import dev.rollczi.litecommands.LiteCommands
import dev.rollczi.litecommands.bukkit.LiteBukkitFactory
import gg.mineral.practice.arena.Arena
import gg.mineral.practice.category.Category
import gg.mineral.practice.commands.config.*
import gg.mineral.practice.commands.context.ProfileContext
import gg.mineral.practice.commands.context.SpawnLocationContext
import gg.mineral.practice.commands.duel.AcceptCommand
import gg.mineral.practice.commands.duel.DuelCommand
import gg.mineral.practice.commands.events.EventCommand
import gg.mineral.practice.commands.kit.LeaveCommand
import gg.mineral.practice.commands.match.ForfeitCommand
import gg.mineral.practice.commands.party.PartyCommand
import gg.mineral.practice.commands.resolver.*
import gg.mineral.practice.commands.settings.*
import gg.mineral.practice.commands.spectator.FollowCommand
import gg.mineral.practice.commands.spectator.SpectateCommand
import gg.mineral.practice.commands.spectator.StopSpectatingCommand
import gg.mineral.practice.commands.stats.EloCommand
import gg.mineral.practice.commands.stats.LeaderboardsCommand
import gg.mineral.practice.commands.stats.PotsCommand
import gg.mineral.practice.commands.stats.ViewInventoryCommand
import gg.mineral.practice.commands.testing.BotTestingCommand
import gg.mineral.practice.commands.tournament.JoinCommand
import gg.mineral.practice.commands.tournament.TournamentCommand
import gg.mineral.practice.entity.Profile
import gg.mineral.practice.entity.ProfileData
import gg.mineral.practice.gametype.Gametype
import gg.mineral.practice.listeners.*
import gg.mineral.practice.managers.ArenaManager
import gg.mineral.practice.managers.CategoryManager
import gg.mineral.practice.managers.GametypeManager
import gg.mineral.practice.managers.QueuetypeManager
import gg.mineral.practice.queue.Queuetype
import gg.mineral.practice.util.world.SpawnLocation
import gg.mineral.practice.util.world.VoidWorldGenerator
import gg.mineral.server.combat.KnockbackProfile
import io.github.retrooper.packetevents.factory.spigot.SpigotPacketEventsBuilder
import org.bukkit.Bukkit
import org.bukkit.command.CommandSender
import org.bukkit.event.Listener
import org.bukkit.plugin.java.JavaPlugin


class PracticePlugin : JavaPlugin() {
    companion object {
        lateinit var INSTANCE: PracticePlugin
    }

    override fun onLoad() {
        INSTANCE = this

        PacketEvents.setAPI(SpigotPacketEventsBuilder.build(this))
        PacketEvents.getAPI().load()
    }

    private lateinit var liteCommands: LiteCommands<CommandSender>

    override fun onEnable() {
        PacketEvents.getAPI().init()

        ArenaManager.load()
        QueuetypeManager.load()
        CategoryManager.load()
        GametypeManager.load()
        // TODO: change invalid command messages
        this.liteCommands = LiteBukkitFactory.builder("Practice", this)
            .commands(
                PracticeCommand(),
                ArenaCommand(),
                CategoryCommand(),
                GametypeCommand(),
                KitEditorCommand(),
                LeaderboardConfigCommand(),
                LobbyCommand(),
                QueuetypeCommand(),
                PartiesCommand(),
                SettingsConfigCommand(),
                SpectateConfigCommand(),
                AcceptCommand(),
                DuelCommand(),
                EventCommand(),
                LeaveCommand(),
                PartyCommand(),
                ToggleDuelRequestsCommand(),
                SettingsCommand(),
                DayCommand(),
                NightCommand(),
                TogglePartyRequestsCommand(),
                TogglePlayerVisibilityCommand(),
                ToggleScoreboardCommand(),
                FollowCommand(),
                SpectateCommand(),
                StopSpectatingCommand(),
                EloCommand(),
                ViewInventoryCommand(),
                PotsCommand(),
                LeaderboardsCommand(),
                TournamentCommand(),
                JoinCommand(),
                BotTestingCommand(),
                ForfeitCommand()
            )
            .context(Profile::class.java, ProfileContext())
            .context(
                SpawnLocation::class.java, SpawnLocationContext()
            )
            .argument(
                Arena::class.java, ArenaResolver()
            )
            .argument(
                Category::class.java, CategoryResolver()
            )
            .argument(
                Gametype::class.java, GametypeResolver()
            )
            .argument(
                KnockbackProfile::class.java, KnockbackProfileResolver()
            )
            .argument(
                Profile::class.java, ProfileResolver()
            )
            .argument(
                ProfileData::class.java, ProfileDataResolver()
            )
            .argument(
                Queuetype::class.java, QueuetypeResolver()
            )
            .build()

        PacketEvents.getAPI().eventManager.registerListener(
            PacketEventsListener(), PacketListenerPriority.NORMAL
        )
        registerListeners(
            BuildListener(), InteractListener(), ComsumeListener(), InventoryListener(),
            DeathListener(), DamageListener(), EntryListener(), HealthListener(),
            MovementListener(), ProjectileListener(), CommandListener()
        )
    }

    private fun initCommands() {

    }

    override fun onDisable() {
        Bukkit.getServer().scheduler.cancelTasks(this)
        PacketEvents.getAPI().terminate()
        this.liteCommands.unregister()
    }

    override fun getDefaultWorldGenerator(worldName: String, id: String) = VoidWorldGenerator()

    private fun registerListeners(vararg listeners: Listener) {
        for (listener in listeners) Bukkit.getPluginManager().registerEvents(listener, this)
    }
}
